pkg/compactor/blocks_cleaner.go

```diff
@@ -419,7 +419,19 @@ func (c *BlocksCleaner) cleanUserPartialBlocks(ctx context.Context, partials map
 		// We can safely delete only partial blocks with a deletion mark.
 		err := metadata.ReadMarker(ctx, userLogger, userBucket, blockID.String(), &metadata.DeletionMark{})
 		if errors.Is(err, metadata.ErrorMarkerNotFound) {
-			return nil
+			//If only visit marker exists in the block, we can safely delete it.
+			isEmpty := true
+			visitMarkerError := userBucket.ReaderWithExpectedErrs(IsNotBlockVisitMarkerError).Iter(ctx, blockID.String(), func(file string) error {
```

Comment:
I am wondering if naming this variable should be name `notVisitMarkerError` ?

Reply:
also will `visitMarkerError` ever be not nil and contains error other than `ErrorNotBlockVisitMarker` type?

Reply:
One suggest I have to make the code more straight forward to understand is this: 

```go
err = userBucket.ReaderWithExpectedErrs(IsStopIterationError).Iter(ctx, blockID.String(), func(file string) error {
   if IsBlockVisitMarker(file) {
       return nil;
   }

   isEmpty = false
   return StopIteration
}

if IsStopIterationError(err) {
  err = nil
}

```

Essentially I think it would be better if we limit the logic of "empty bucket" inside the call back function to the `Iter` function. Outside of the callback function the exact definition of being empty should be abstracted away.

Reply:
I had the same comment. Seems the intention to return an error is to "fail fast" and stop iterating on the files.

https://github.com/thanos-io/objstore/blob/d1711f1d29240a3276b024577a509d4da39c47f1/objstore.go#L483-L494

Reply:
@alvinlin123 This was what I had some dicussion with @alanprot offline. `Iter` would iterate over files in alphabetical order. So, if a block got tons of chunks file without meta file, the straight forward code would iterate over all chunks file then decide to skip deleting this partial block. Whereas returning error when first chunk file got discovered, it would skip going through the rest of files then skip deletion.

Reply:
@alvinlin123 I renamed the variable and simplified the code a little bit. Please take another look.

Reply:
I still feel that havig `ErrorNotBlockVisitMarker` is quite misleading. However, it may be just me. I will ask others to chime in as well

@friedrichg may I kindly ask you to review this PR?

Reply:
@alanprot Added comments